csharp/src/Ice/Proxy.cs

```diff
@@ -322,76 +325,170 @@ public static class Proxy
+                        catch (InvalidRequestHandlerException)
                         {
-                            // Clear the proxy's cached request handler if connection caching is enabled
+                            // TODO temporary, this is already fixed in the transport refactoring branch
+                            // Not a retry, we try to send the request using a handler that was no longer valid
                             if (reference.IsConnectionCached)
                             {
-                                proxy.IceReference.ClearRequestHandler(handler!);
+                                reference.ClearRequestHandler(handler!);
+                            }
+                            continue;
+                        }
+                        catch (NoMoreReplicasException)
+                        {
+                            // The last exception asked to retry using a different replica, but we already tried all
+                            // known replicas.
+                            if (lastException != null)
+                            {
+                                throw ExceptionUtil.Throw(lastException);
+                            }
+                            else
+                            {
+                                Debug.Assert(response != null && response.ResultType == ResultType.Failure);
+                                return response;
                             }
                         }
-                        catch (OperationCanceledException)
+                        catch (ConnectionClosedByPeerException ex)
                         {
-                            throw; // Don't retry cancelled operations
+                            // Always retry a gracefully close connection. Don't exclude the connector after a
+                            // graceful close connection, in case it is the only connector available.
+                            lastException = ex;
+                            retryable = Retryable.AfterDelay;
+                        }
+                        catch (TransportException ex) when (request.IsIdempotent || !progressWrapper.IsSent)
+                        {
+                            // Retry transport exceptions if the request is idempotent or was not send
+                            lastException = ex;
+                            retryable = ex.Retryable;
+                        }
+
+                        if (response != null)
+                        {
+                            Debug.Assert(response.ResultType == ResultType.Failure);
+                            observer?.RemoteException();
+                            if (response.ReadSystemException(proxy.Communicator) is Exception systemException)
+                            {
+                                // 1.1 system exceptions
+                                if (systemException is ObjectNotExistException && reference.IsWellKnown)
```

Comment:
We don't only retry ONE for well-known with 3.7, why only keep well-known here? Don't we also need to add the ice_add_proxy quirk here?

Reply:
fixed